File: app/services/tasks.py
"""Background tasks — one system for every user-triggered long operation.

Generalized from `intake/jobs.py`, which was the only asynchronous path in the
product. Everything else either grew its own job concept or blocked the HTTP
request end to end.

Backed by `assessment_job` with a `kind` discriminator (migration 0049).
`assessment_id` stays intake's column and is simply null for other kinds; a
non-intake task reports what it produced through `result.result_id`.

`job_run` (scheduled cron work) is NOT part of this. Cron jobs have nobody
waiting on them, and a progress bar nobody is watching is a lie about what the
tracker is for.

Rule this module exists to keep: **a task that outlives a reader's patience
still reports honestly.** A stage is only ever set from real progress; nothing
here advances a stage on a timer.
"""
from __future__ import annotations

import logging
from datetime import datetime, timezone
from uuid import uuid4

from app.db import supabase_rest_get, supabase_rest_patch, supabase_rest_post

log = logging.getLogger(__name__)

TABLE = "assessment_job"

#: Kinds this module accepts. Mirrored on the client in web/src/jobs/tasks.ts —
#: the two lists must agree or the tracker cannot resolve a status URL.
KINDS = ("intake", "reassessment", "quarterly_build", "report_pdf")

#: Ordered stages per kind. Mirrored client-side ONLY to compute progress; an
#: unrecognised stage renders indeterminate there rather than guessing a
#: percentage (Hard Rule 7). Adding a stage here means adding it there too.
STAGES: dict[str, tuple[str, ...]] = {
    "intake": (
        "queued", "fetching", "extracting", "segmenting", "classifying",
        "profiling", "benchmarking", "scoring", "generating_findings",
    ),
    "reassessment": ("queued", "loading_notices", "scoring", "writing_snapshots"),
    "quarterly_build": ("queued", "aggregating", "suppressing", "rendering"),
    "report_pdf": ("queued", "rendering"),
}


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


async def create(
    *,
    kind: str,
    organization_id: str | None = None,
    created_by: str | None = None,
    idempotency_key: str | None = None,
) -> dict:
    """Create a queued task.

    On an idempotency-key race (unique-index conflict) the EXISTING task is
    returned rather than a duplicate created — a double-submit must never start
    the same expensive work twice.
    """
    if kind not in KINDS:
        raise ValueError(f"unknown task kind: {kind!r}")

    task_id = str(uuid4())
    payload = {
        "job_id": task_id,
        "kind": kind,
        "organization_id": organization_id,
        "created_by": created_by,
        "idempotency_key": idempotency_key,
        "status": "queued",
        "stage": "queued",
        "created_at": _now(),
        "updated_at": _now(),
    }
    r = await supabase_rest_post(TABLE, payload)
    if r.status_code == 409 and idempotency_key:
        existing = await find_by_idempotency_key(idempotency_key)
        if existing:
            return {**existing, "_idempotent_replay": True}
    if r.status_code >= 400:
        raise RuntimeError(f"could not create {TABLE} (HTTP {r.status_code})")
    return payload


async def find_by_idempotency_key(key: str | None) -> dict | None:
    if not key:
        return None
    r = await supabase_rest_get(
        TABLE, select="*", filters=f"idempotency_key=eq.{key}", limit=1)
    rows = r.json() if r.status_code == 200 else []
    return rows[0] if rows else None


async def set_stage(task_id: str, stage: str, *, status: str = "running") -> None:
    await supabase_rest_patch(
        TABLE, f"job_id=eq.{task_id}",
        {"stage": stage, "status": status, "updated_at": _now()})


async def complete(task_id: str, *, result: dict, result_id: str | None = None) -> None:
    """Mark a task complete.

    `result_id` is what the reader can open next — a snapshot id, a report id.
    It is stored inside `result` so the shape stays kind-agnostic; only intake
    also writes the dedicated `assessment_id` column, which predates this module.
    """
    patch = {
        "status": "complete", "stage": "complete",
        "result": {**result, "result_id": result_id} if result_id else result,
        "updated_at": _now(),
    }
    await supabase_rest_patch(TABLE, f"job_id=eq.{task_id}", patch)


async def fail(task_id: str, error: str) -> None:
    await supabase_rest_patch(
        TABLE, f"job_id=eq.{task_id}",
        {"status": "failed", "stage": "failed", "error": (error or "")[:2000],
         "updated_at": _now()})


async def get(task_id: str) -> dict | None:
    r = await supabase_rest_get(
        TABLE, select="*", filters=f"job_id=eq.{task_id}", limit=1)
    rows = r.json() if r.status_code == 200 else []
    return rows[0] if rows else None


def to_status(row: dict) -> dict:
    """Shape a row into the status payload the tracker polls.

    Deliberately narrow: status, stage, an optional openable id, and an error.
    The tracker must not be able to read anything it should not, and a status
    endpoint is the easiest place to leak a whole row by accident.
    """
    result = row.get("result") or {}
    return {
        "status": row.get("status") or "queued",
        "stage": row.get("stage") or "queued",
        "kind": row.get("kind") or "intake",
        "result_id": result.get("result_id") or row.get("assessment_id"),
        "error": row.get("error"),
    }

File: web/src/jobs/tasks.ts
/**
 * Background-task model — shared by the provider and the tracker widget.
 *
 * ONE task system, not one per feature. Intake was asynchronous first and is the
 * pattern generalized here; re-assessment, quarterly build and report export
 * join it rather than each growing their own progress affordance. A reader who
 * has to look in three places to find out what is running has lost track of it.
 *
 * Stage lists are SERVER-authoritative. They are mirrored here only to turn a
 * reported stage into "how far along am I", and the mirror is deliberately
 * narrow: an unrecognised stage leaves progress INDETERMINATE rather than
 * guessing a percentage — a fabricated 40% is a fabricated number (Hard Rule 7,
 * F01 AC-15).
 */

export type TaskKind =
  | "intake"
  | "reassessment"
  | "quarterly_build"
  | "report_pdf";

export type TaskStatus = "queued" | "running" | "ready" | "failed";

export interface Task {
  taskId: string;
  kind: TaskKind;
  /** What the reader asked for, in their words. Shown in the tracker. */
  label: string;
  status: TaskStatus;
  stage: string;
  submittedAt: number;
  /** Set when the task produces something openable. */
  resultId?: string;
  /** Where "open" goes when it finishes. Absent = nothing to open. */
  resultHref?: string;
  error?: string;
  dismissed?: boolean;
}

/** Status endpoint per kind. Kept beside the kinds so adding one is one edit. */
export const TASK_STATUS_URL: Record<TaskKind, (id: string) => string> = {
  intake:          id => `/assessments/${id}/status`,
  reassessment:    id => `/tasks/${id}`,
  quarterly_build: id => `/tasks/${id}`,
  report_pdf:      id => `/tasks/${id}`,
};

/** Ordered stages per kind, for the progress bar. */
export const TASK_STAGES: Record<TaskKind, readonly string[]> = {
  intake: [
    "queued", "fetching", "extracting", "segmenting", "classifying",
    "profiling", "benchmarking", "scoring", "generating_findings",
  ],
  reassessment:    ["queued", "loading_notices", "scoring", "writing_snapshots"],
  quarterly_build: ["queued", "aggregating", "suppressing", "rendering"],
  report_pdf:      ["queued", "rendering"],
};

export const TERMINAL_STAGES = ["complete", "failed"] as const;

export const STAGE_LABELS: Record<string, string> = {
  queued: "Queued",
  // intake
  fetching: "Fetching the notice",
  extracting: "Extracting text",
  segmenting: "Decomposing into clauses",
  classifying: "Classifying clauses",
  profiling: "Profiling the organization",
  benchmarking: "Building the peer benchmark",
  scoring: "Scoring against peers",
  generating_findings: "Generating findings",
  awaiting_review: "Awaiting expert review",
  generating_report: "Generating the report",
  // reassessment
  loading_notices: "Loading stored notices",
  writing_snapshots: "Writing new snapshots",
  // quarterly
  aggregating: "Aggregating the corpus",
  suppressing: "Applying suppression rules",
  rendering: "Rendering",
  // terminal
  complete: "Complete",
  failed: "Could not be completed",
  still_running: "Still processing",
};

export function stageLabel(stage: string): string {
  return STAGE_LABELS[stage] ?? stage.replace(/_/g, " ");
}

/**
 * Fraction complete, or `null` when the stage is not on the known list.
 * Null renders an indeterminate bar — never an invented percentage.
 */
export function stageProgress(kind: TaskKind, stage: string): number | null {
  // Complete is the only 1.0. Everything mid-pipeline stays below it, so a bar
  // can never read "done" while work is still running.
  if (stage === "complete") return 1;
  const stages = TASK_STAGES[kind];
  if (!stages) return null;
  const i = stages.indexOf(stage);
  if (i < 0) return null;
  return (i + 1) / (stages.length + 1);
}

export function isTerminal(task: Pick<Task, "status" | "stage">): boolean {
  return task.status === "ready" || task.status === "failed" ||
         (TERMINAL_STAGES as readonly string[]).includes(task.stage);
}
